File: qlat/pylib/qlat/qcd.py
import qlat.cqlat as c

from qlat_utils import *

from qlat.field import *
from qlat.field_utils import *
from qlat.utils_io import *

class GaugeField(Field):

    def __init__(self, geo = None, *, ctype = None, multiplicity = None):
        Field.__init__(self, "ColorMatrix", geo, 4)

    def copy(self, is_copying_data = True):
        f = GaugeField()
        if is_copying_data:
            f @= self
        return f

    @timer
    def save(self, path):
        mk_file_dirs_info(path)
        return c.save_gauge_field(self, path)

    @timer
    def load(self, path):
        return c.load_gauge_field(self, path)

    def set_rand(self, rng, sigma = 0.5, n_step = 1):
        set_g_rand_color_matrix_field(self, rng, sigma, n_step)

    def unitarize(self):
        c.unitarize_color_matrix_field(self)

    def plaq(self):
        return gf_avg_plaq(self)

    def link_trace(self):
        return gf_avg_link_trace(self)

    def twist_boundary_at_boundary(self, lmom : float = -0.5, mu : int = 3):
        # modify in place
        gf_twist_boundary_at_boundary(self, lmom, mu)

    def show_info(self):
        gf_show_info(self)

###

class GaugeTransform(Field):

    def __init__(self, geo = None, *, ctype = None, multiplicity = None):
        Field.__init__(self, "ColorMatrix", geo, 1)

    def copy(self, is_copying_data = True):
        f = GaugeTransform()
        if is_copying_data:
            f @= self
        return f

    def set_rand(self, rng, sigma = 0.5, n_step = 1):
        set_g_rand_color_matrix_field(self, rng, sigma, n_step)

    def unitarize(self):
        c.unitarize_color_matrix_field(self)

    def __mul__(self, other):
        # other can be GaugeTransform, GaugeField, Prop, SelProp, PselProp, list
        from qlat.propagator import Prop, SelProp, PselProp
        if isinstance(other, GaugeTransform):
            gt = GaugeTransform()
            c.apply_gt_gt(gt, self, other)
            return gt
        elif isinstance(other, GaugeField):
            gf = GaugeField()
            c.apply_gt_gf(gf, self, other)
            return gf
        elif isinstance(other, Prop):
            prop = Prop()
            c.apply_gt_prop(prop, self, other)
            return prop
        elif isinstance(other, SelProp):
            prop = SelProp(other.fsel)
            c.apply_gt_sprop(prop, self, other)
            return prop
        elif isinstance(other, PselProp):
            prop = PselProp(other.psel)
            c.apply_gt_psprop(prop, self, other)
            return prop
        elif isinstance(other, list):
            return [ self * p for p in other ]
        else:
            raise Exception("GaugeTransform.__mul__")

    def inv(self):
        gt = GaugeTransform()
        c.gt_invert(gt, self)
        return gt

###

@timer_verbose
def gf_show_info(gf):
    assert isinstance(gf, GaugeField)
    displayln_info(f"gf_show_info: plaq = {gf.plaq():.16F} ; link_trace = {gf.link_trace():.16F}.")

def gf_avg_plaq(gf):
    assert isinstance(gf, GaugeField)
    return c.gf_avg_plaq(gf)

def gf_avg_link_trace(gf):
    assert isinstance(gf, GaugeField)
    return c.gf_avg_link_trace(gf)

def gf_wilson_line_no_comm(wlf, m, gf_ext, path, path_n = None):
    # wlf = Field("ColorMatrix", geo)
    # will only modify the m'th component
    # e.g. path = [ mu, mu, nu, -mu-1, -mu-1, ]
    # e.g. path = [ mu, nu, -mu-1, ], path_n = [ 2, 1, 2, ]
    if path_n is None:
        c.gf_wilson_line_no_comm(wlf, m, gf_ext, path)
    else:
        c.gf_wilson_line_no_comm(wlf, m, gf_ext, path, path_n)

def gf_wilson_lines_no_comm(gf_ext, path_list):
    # path_list = [ path_spec, ... ]
    # e.g. path_spec = [ mu, mu, nu, -mu-1, -mu-1, ]
    # e.g. path_spec = ([ mu, nu, -mu-1, ], [ 2, 1, 2, ],)
    # return wlf
    multiplicity = len(path_list)
    geo = geo_reform(gf_ext.geo(), multiplicity)
    wlf = Field("ColorMatrix", geo)
    for m, p in enumerate(path_list):
        if isinstance(p, tuple) and len(p) == 2:
            path, path_n = p
            gf_wilson_line_no_comm(wlf, m, gf_ext, path, path_n)
        else:
            path = p
            gf_wilson_line_no_comm(wlf, m, gf_ext, path)
    return wlf

def gf_avg_wilson_loop_normalized_tr(gf, l, t):
    assert isinstance(gf, GaugeField)
    assert isinstance(l, int)
    assert isinstance(t, int)
    return c.gf_avg_wilson_loop_normalized_tr(gf, l, t)

def set_g_rand_color_matrix_field(fc, rng, sigma, n_steps = 1):
    assert isinstance(fc, Field) and fc.ctype == "ColorMatrix"
    assert isinstance(rng, RngState)
    return c.set_g_rand_color_matrix_field(fc, rng, sigma, n_steps)

def gf_twist_boundary_at_boundary(gf : GaugeField, lmom : float = -0.5, mu : int = 3):
    # modify gf in place
    c.gf_twist_boundary_at_boundary(gf, lmom, mu)

def mk_left_expanded_gauge_field(gf):
    gf1 = field_expanded(gf, 1, 0)
    refresh_expanded_1(gf1)
    return gf1

File: qlat/pylib/cqlat/qcd.cpp
#include "lib.h"

EXPORT(gf_show_info, {
  using namespace qlat;
  PyObject* p_gf = NULL;
  if (!PyArg_ParseTuple(args, "O", &p_gf)) {
    return NULL;
  }
  const GaugeField& gf = py_convert_type<GaugeField>(p_gf);
  gf_show_info(gf);
  Py_RETURN_NONE;
})

EXPORT(gf_avg_plaq, {
  using namespace qlat;
  PyObject* p_gf = NULL;
  if (!PyArg_ParseTuple(args, "O", &p_gf)) {
    return NULL;
  }
  const GaugeField& gf = py_convert_type<GaugeField>(p_gf);
  const double ret = gf_avg_plaq(gf);
  return py_convert(ret);
})

EXPORT(gf_avg_link_trace, {
  using namespace qlat;
  PyObject* p_gf = NULL;
  if (!PyArg_ParseTuple(args, "O", &p_gf)) {
    return NULL;
  }
  const GaugeField& gf = py_convert_type<GaugeField>(p_gf);
  const double ret = gf_avg_link_trace(gf);
  return py_convert(ret);
})

EXPORT(gf_avg_wilson_loop_normalized_tr, {
  using namespace qlat;
  PyObject* p_gf = NULL;
  int l = 0;
  int t = 0;
  if (!PyArg_ParseTuple(args, "Oii", &p_gf, &l, &t)) {
    return NULL;
  }
  const GaugeField& gf = py_convert_type<GaugeField>(p_gf);
  const double ret = matrix_trace(gf_avg_wilson_loop(gf, l, t)).real() / 3.0;
  return py_convert(ret);
})

EXPORT(gf_wilson_line_no_comm, {
  using namespace qlat;
  PyObject* p_wilson_line_field = NULL;
  int wilson_line_field_m = 0;
  PyObject* p_gf_ext = NULL;
  PyObject* p_path = NULL;
  PyObject* p_path_n = NULL;
  if (!PyArg_ParseTuple(args, "OiOO|O", &p_wilson_line_field,
                        &wilson_line_field_m, &p_gf_ext, &p_path, &p_path_n)) {
    return NULL;
  }
  Field<ColorMatrix>& wilson_line_field =
      py_convert_type_field<ColorMatrix>(p_wilson_line_field);
  const GaugeField& gf_ext = py_convert_type<GaugeField>(p_gf_ext);
  const std::vector<int> path = py_convert_data<std::vector<int> >(p_path);
  if (p_path_n != NULL) {
    const std::vector<int> path_n =
        py_convert_data<std::vector<int> >(p_path_n);
    gf_wilson_line_no_comm(wilson_line_field, wilson_line_field_m, gf_ext, path,
                           path_n);
  } else {
    gf_wilson_line_no_comm(wilson_line_field, wilson_line_field_m, gf_ext,
                           path);
  }
  Py_RETURN_NONE;
})

EXPORT(set_g_rand_color_matrix_field, {
  using namespace qlat;
  PyObject* p_field = NULL;
  PyObject* p_rng = NULL;
  double sigma = 1.0;
  int n_step = 1;
  if (!PyArg_ParseTuple(args, "OOd|i", &p_field, &p_rng, &sigma, &n_step)) {
    return NULL;
  }
  Field<ColorMatrix>& field = py_convert_type<Field<ColorMatrix> >(p_field);
  const RngState& rng = py_convert_type<RngState>(p_rng);
  set_g_rand_color_matrix_field(field, rng, sigma, n_step);
  Py_RETURN_NONE;
})

EXPORT(unitarize_color_matrix_field, {
  using namespace qlat;
  PyObject* p_field = NULL;
  if (!PyArg_ParseTuple(args, "O", &p_field)) {
    return NULL;
  }
  Field<ColorMatrix>& field = py_convert_type<Field<ColorMatrix> >(p_field);
  unitarize(field);
  Py_RETURN_NONE;
})

EXPORT(save_gauge_field, {
  using namespace qlat;
  PyObject* p_gf = NULL;
  PyObject* p_path = NULL;
  if (!PyArg_ParseTuple(args, "OO", &p_gf, &p_path)) {
    return NULL;
  }
  const GaugeField& gf = py_convert_type<GaugeField>(p_gf);
  std::string path;
  py_convert(path, p_path);
  const long ret = save_gauge_field(gf, path);
  return py_convert(ret);
})

EXPORT(load_gauge_field, {
  using namespace qlat;
  PyObject* p_gf = NULL;
  PyObject* p_path = NULL;
  if (!PyArg_ParseTuple(args, "OO", &p_gf, &p_path)) {
    return NULL;
  }
  GaugeField& gf = py_convert_type<GaugeField>(p_gf);
  const std::string path = py_convert_data<std::string>(p_path);
  const long ret = load_gauge_field(gf, path);
  return py_convert(ret);
})

EXPORT(gf_twist_boundary_at_boundary, {
  using namespace qlat;
  PyObject* p_gf = NULL;
  double lmom = 0.0;
  int mu = -1;
  if (!PyArg_ParseTuple(args, "Odi", &p_gf, &lmom, &mu)) {
    return NULL;
  }
  GaugeField& gf = py_convert_type<GaugeField>(p_gf);
  qassert(0 <= mu and mu < 4);
  twist_boundary_at_boundary(gf, lmom, mu);
  Py_RETURN_NONE;
})

EXPORT(save_gauge_transform_cps, {
  using namespace qlat;
  PyObject* p_gt = NULL;
  PyObject* p_path = NULL;
  if (!PyArg_ParseTuple(args, "OO", &p_gt, &p_path)) {
    return NULL;
  }
  const GaugeTransform& gt = py_convert_type<GaugeTransform>(p_gt);
  const std::string path = py_convert_data<std::string>(p_path);
  const long ret = save_gauge_transform_cps(gt, path);
  return py_convert(ret);
})

EXPORT(load_gauge_transform_cps, {
  using namespace qlat;
  PyObject* p_gt = NULL;
  PyObject* p_path = NULL;
  if (!PyArg_ParseTuple(args, "OO", &p_gt, &p_path)) {
    return NULL;
  }
  GaugeTransform& gt = py_convert_type<GaugeTransform>(p_gt);
  const std::string path = py_convert_data<std::string>(p_path);
  const long ret = load_gauge_transform_cps(gt, path);
  return py_convert(ret);
})

EXPORT(apply_gt_gt, {
  using namespace qlat;
  PyObject* p_gt = NULL;
  PyObject* p_gt1 = NULL;
  PyObject* p_gt0 = NULL;
  if (!PyArg_ParseTuple(args, "OOO", &p_gt, &p_gt1, &p_gt0)) {
    return NULL;
  }
  // p_gt <- p_gt1 * p_gt0
  GaugeTransform& gt = py_convert_type<GaugeTransform>(p_gt);
  const GaugeTransform& gt1 = py_convert_type<GaugeTransform>(p_gt1);
  const GaugeTransform& gt0 = py_convert_type<GaugeTransform>(p_gt0);
  gt_apply_gauge_transformation(gt, gt0, gt1);
  Py_RETURN_NONE;
})

EXPORT(apply_gt_gf, {
  using namespace qlat;
  PyObject* p_gf = NULL;
  PyObject* p_gt = NULL;
  PyObject* p_gf0 = NULL;
  if (!PyArg_ParseTuple(args, "OOO", &p_gf, &p_gt, &p_gf0)) {
    return NULL;
  }
  // p_gf <- p_gt * p_gf0
  GaugeField& gf = py_convert_type<GaugeField>(p_gf);
  const GaugeTransform& gt = py_convert_type<GaugeTransform>(p_gt);
  const GaugeField& gf0 = py_convert_type<GaugeField>(p_gf0);
  gf_apply_gauge_transformation(gf, gf0, gt);
  Py_RETURN_NONE;
})

EXPORT(apply_gt_prop, {
  using namespace qlat;
  PyObject* p_prop = NULL;
  PyObject* p_gt = NULL;
  PyObject* p_prop0 = NULL;
  if (!PyArg_ParseTuple(args, "OOO", &p_prop, &p_gt, &p_prop0)) {
    return NULL;
  }
  // p_prop <- p_gt * p_prop0
  Propagator4d& prop = py_convert_type<Propagator4d>(p_prop);
  const GaugeTransform& gt = py_convert_type<GaugeTransform>(p_gt);
  const Propagator4d& prop0 = py_convert_type<Propagator4d>(p_prop0);
  prop_apply_gauge_transformation(prop, prop0, gt);
  Py_RETURN_NONE;
})

EXPORT(apply_gt_sprop, {
  using namespace qlat;
  PyObject* p_prop = NULL;
  PyObject* p_gt = NULL;
  PyObject* p_prop0 = NULL;
  if (!PyArg_ParseTuple(args, "OOO", &p_prop, &p_gt, &p_prop0)) {
    return NULL;
  }
  // p_prop <- p_gt * p_prop0
  SelectedField<WilsonMatrix>& prop =
      py_convert_type<SelectedField<WilsonMatrix> >(p_prop);
  const GaugeTransform& gt = py_convert_type<GaugeTransform>(p_gt);
  const SelectedField<WilsonMatrix>& prop0 =
      py_convert_type<SelectedField<WilsonMatrix> >(p_prop0);
  const FieldSelection& fsel = py_convert_type<FieldSelection>(p_prop0, "fsel");
  prop_apply_gauge_transformation(prop, prop0, gt, fsel);
  Py_RETURN_NONE;
})

EXPORT(apply_gt_psprop, {
  using namespace qlat;
  PyObject* p_prop = NULL;
  PyObject* p_gt = NULL;
  PyObject* p_prop0 = NULL;
  if (!PyArg_ParseTuple(args, "OOO", &p_prop, &p_gt, &p_prop0)) {
    return NULL;
  }
  // p_prop <- p_gt * p_prop0
  SelectedPoints<WilsonMatrix>& prop =
      py_convert_type<SelectedPoints<WilsonMatrix> >(p_prop);
  const GaugeTransform& gt = py_convert_type<GaugeTransform>(p_gt);
  const SelectedPoints<WilsonMatrix>& prop0 =
      py_convert_type<SelectedPoints<WilsonMatrix> >(p_prop0);
  const PointSelection& psel = py_convert_type<PointSelection>(p_prop0, "psel");
  prop_apply_gauge_transformation(prop, prop0, gt, psel);
  Py_RETURN_NONE;
})

EXPORT(gt_invert, {
  using namespace qlat;
  PyObject* p_gt = NULL;
  PyObject* p_gt0 = NULL;
  if (!PyArg_ParseTuple(args, "OO", &p_gt, &p_gt0)) {
    return NULL;
  }
  // p_gt <- p_gt0^{-1}
  GaugeTransform& gt = py_convert_type<GaugeTransform>(p_gt);
  const GaugeTransform& gt0 = py_convert_type<GaugeTransform>(p_gt0);
  gt_invert(gt, gt0);
  Py_RETURN_NONE;
})
